File: src/main/java/net/punchtree/loquainteractable/input/SteerVehicleInputPacketAdapter.java
package net.punchtree.loquainteractable.input;

import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;

import com.comphenix.protocol.PacketType;
import com.comphenix.protocol.ProtocolManager;
import com.comphenix.protocol.events.ListenerPriority;
import com.comphenix.protocol.events.PacketAdapter;
import com.comphenix.protocol.events.PacketContainer;
import com.comphenix.protocol.events.PacketEvent;

public class SteerVehicleInputPacketAdapter extends PacketAdapter {

	private final ProtocolManager protocolManager;
	private final PlayerInputsManager playerInputsManager;
	
	public SteerVehicleInputPacketAdapter(Plugin plugin, ProtocolManager protocolManager, PlayerInputsManager playerInputsManager) {
		super(plugin, ListenerPriority.LOWEST, new PacketType[] { PacketType.Play.Client.STEER_VEHICLE });
		this.protocolManager = protocolManager;
		this.playerInputsManager = playerInputsManager;
	}
	
	public void handleInput() {
		protocolManager.addPacketListener(this);
	}
		
	public void stopHandlingInput() {
		protocolManager.removePacketListener(this);
	}
	
	@Override
	public void onPacketReceiving(PacketEvent packetEvent) {
		
		PacketContainer packetContainer = packetEvent.getPacket();
		
		float sidewaysAxis = packetContainer.getFloat().readSafely(0).floatValue();
		float forwardAxis = packetContainer.getFloat().readSafely(1).floatValue();
		boolean jump = packetContainer.getBooleans().readSafely(0);
		boolean unmount = packetContainer.getBooleans().readSafely(1);

		Player player = packetEvent.getPlayer();
		
		PlayerInputs playerInputs = playerInputsManager.getInputsForPlayer(player);
		assert(playerInputs != null);
		playerInputs.setVehicleForwardAxis(forwardAxis);
		playerInputs.setVehicleSidewaysAxis(sidewaysAxis);
		playerInputs.setVehicleJump(jump);
		playerInputs.setVehicleUnmount(unmount);
	}
		
	
	
}

File: src/main/java/net/punchtree/loquainteractable/input/SteerVehicleInputPacketAdapter.kt
package net.punchtree.loquainteractable.input

import com.comphenix.protocol.PacketType
import com.comphenix.protocol.events.ListenerPriority
import com.comphenix.protocol.events.PacketAdapter
import com.comphenix.protocol.events.PacketEvent
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.world.entity.player.Input
import org.bukkit.plugin.Plugin

class SteerVehicleInputPacketAdapter(
    plugin: Plugin,
    private val playerInputsManager: PlayerInputsManager
) : PacketAdapter(
    plugin,
    ListenerPriority.LOWEST,
    *arrayOf(
        PacketType.Play.Client.STEER_VEHICLE
    )
) {
    override fun onPacketReceiving(packetEvent: PacketEvent) {
        val packetContainer = packetEvent.packet

        packetContainer.serializeToBuffer()

        val buffer = packetContainer.serializeToBuffer() as FriendlyByteBuf
        val input = Input.STREAM_CODEC.decode(buffer)

        val player = packetEvent.player

        val playerInputs = playerInputsManager.getInputsForPlayer(player)
        playerInputs?.updateVehicleInput(player, input)
    }
}
